kubernetes/mutating-admission/main.rego

```diff
@@ -0,0 +1,179 @@
+package library.kubernetes.admission.mutating
+
+###########################################################################
+# Implementation of the k8s admission control external webhook interface,
+# combining validating and mutating admission controllers
+###########################################################################
+
+main = {
+	"apiVersion": "admission.k8s.io/v1beta1",
+	"kind": "AdmissionReview",
+	"response": response,
+}
+
+default response = {"allowed": true}
+
+# non-patch response i.e. validation response
+response = x {
+	count(patch) = 0
+
+	x := {
+		"allowed": false,
+		"status": {"reason": reason},
+	}
+
+	reason = concat(", ", deny)
+	reason != ""
+}
+
+# patch response i.e. mutating respone
+response = x {
+	count(patch) > 0
+
+	# if there are missing leaves e.g. trying to add a label to something that doesn't
+	# yet have any, we need to create the leaf nodes as well
+
+	fullPatches := ensureParentPathsExist(cast_array(patch))
+
+	x := {
+		"allowed": true,
+		"patchType": "JSONPatch",
+		"patch": base64url.encode(json.marshal(fullPatches)),
+	}
+}
+
+isValidRequest {
+	# not sure if this might be a race condition, it might get called before 
+	# all the validation rules have been run
+	count(deny) = 0
+}
+
+isCreateOrUpdate {
+	isCreate
+}
+
+isCreateOrUpdate {
+	isUpdate
+}
+
+isCreate {
+	input.request.operation == "CREATE"
+}
+
+isUpdate {
+	input.request.operation == "UPDATE"
+}
+
+###########################################################################
+# PATCH helpers 
+# Note: These rules assume that the input is an object
+# not an AdmissionRequest, because labels and annotations 
+# can apply to various sub-objects within a request
+# So from the context of an AdmissionRequest they need to
+# be called like
+#   hasLabelValue[["foo", "bar"]] with input as input.request.object
+# or
+#   hasLabelValue[["foo", "bar"]] with input as input.request.oldObject
+###########################################################################
+
+hasLabels {
+	input.metadata.labels
+}
+
+hasLabel[label] {
+	hasLabels
```

Comment:
Nit: this line (`hasLabels`) is unnecessary. The following line will fail if labels is missing (i.e., the reference `input.metadata.labels` will be undefined.

Reply:
Yep. Tim actually pointed that out on Slack but I wasn't quite sure what he was referring to. Will remove.